libgnucash/engine/qof-string-cache.cpp

```diff
@@ -43,98 +41,60 @@ static QofLogModule log_module = QOF_MOD_UTIL;
 /* =================================================================== */
 /* The QOF string cache                                                */
 /*                                                                     */
-/* The cache is a GHashTable where a copy of the string is the key,    */
+/* The cache is an std:unordered_map where an std::string is the key,  */
 /* and a ref count is the value                                        */
 /* =================================================================== */
 
-static GHashTable* qof_string_cache = NULL;
-
-static GHashTable*
-qof_get_string_cache(void)
-{
-    if (!qof_string_cache)
-    {
-        qof_string_cache = g_hash_table_new_full(
-                               g_str_hash,               /* hash_func          */
-                               g_str_equal,              /* key_equal_func     */
-                               g_free,                   /* key_destroy_func   */
-                               g_free);                  /* value_destroy_func */
-    }
-    return qof_string_cache;
-}
+static std::unordered_map<std::string, int> cache;
 
 void
 qof_string_cache_init(void)
 {
-    (void)qof_get_string_cache();
+    // nop
 }
 
 void
 qof_string_cache_destroy (void)
 {
-    if (qof_string_cache)
-    {
-        g_hash_table_destroy(qof_string_cache);
-    }
-    qof_string_cache = NULL;
+    cache.clear ();
 }
 
 /* If the key exists in the cache, check the refcount.  If 1, just
  * remove the key.  Otherwise, decrement the refcount */
 void
 qof_string_cache_remove(const char * key)
 {
-    if (key && key[0] != 0)
-    {
-        GHashTable* cache = qof_get_string_cache();
-        gpointer value;
-        gpointer cache_key;
-        if (g_hash_table_lookup_extended(cache, key, &cache_key, &value))
-        {
-            guint* refcount = (guint*)value;
-            if (*refcount == 1)
-            {
-                g_hash_table_remove(cache, key);
-            }
-            else
-            {
-                --(*refcount);
-            }
-        }
-    }
+    if (!key || !key[0])
+        return;
+
+    auto map_iter = cache.find (key);
+    if (map_iter == cache.end())
+        return;
+
+    if (map_iter->second == 1)
+        cache.erase (key);
+    else
+        map_iter->second--;
 }
 
 /* If the key exists in the cache, increment the refcount.  Otherwise,
  * add it with a refcount of 1. */
 const char *
 qof_string_cache_insert(const char * key)
 {
-    if (key)
-    {
-        if (key[0] == 0)
-        {
-            return "";
-        }
+    if (!key)
+        return nullptr;
+
+    if (!key[0])
+        return "";
+
+    auto map_iter = cache.find (key);
+    if (map_iter == cache.end())
+        map_iter = cache.insert_or_assign (key, 1).first;
```

Comment:
You've already established that `key` isn't present so there's no point in `insert_or_assign`, just use `emplace`.